flag/pflag.go

```diff
@@ -0,0 +1,365 @@
+		case reflect.Slice, reflect.Map:
+			switch ft {
+			case stringSlice:
+				f = s.Flags.StringSliceP(name, shorthand, fieldVal.Interface().([]string), help)
+			case mapStringStringSlice:
+				f = s.Flags.VarPF(&mapStringStringSliceFlag{fieldVal.Addr().Interface().(*map[string][]string)}, name, shorthand, help)
+			case mapStringString:
+				f = s.Flags.VarPF(&mapStringStringFlag{fieldVal.Addr().Interface().(*map[string]string)}, name, shorthand, help)
+			case stringSet:
+				f = s.Flags.VarPF(&stringSetFlag{fieldVal.Addr().Interface().(*map[string]struct{})}, name, shorthand, help)
+			default:
+				return fmt.Errorf("unhandled type %s", ft)
+			}
+
+		default:
+			return fmt.Errorf("unhandled type %s", ft)
+		}
+
+		s.flagValues[name] = f
+	}
+	return nil
+}
+
+// Value fills in the user-provided config struct using flags. It looks up the
+// flags to bind into a given struct field by using that field's `dialspflag`
+// struct tag if present, then its `dials` tag if present, and finally its name.
+// If the struct has nested fields, Value will flatten the fields so flags can
+// be defined for nested fields.
+func (s *PflagSet) Value(t *dials.Type) (reflect.Value, error) {
+	// Check whether we've gone through the exercise of parsing flags yet
+	// (and types are compatible).
+	if s.ptrType != nil {
+		if !t.Type().ConvertibleTo(s.ptrType) {
+			return reflect.Value{}, fmt.Errorf(
+				"incompatible types called with Value() (%s) and constructor for flag Source (%s)",
+				t.Type(), s.ptrType)
+		}
+	}
+
+	if s.flagFieldName == nil {
+		s.flagFieldName = map[string]string{}
+	}
+
+	if s.flagValues == nil {
+		s.flagValues = map[string]interface{}{}
+	}
+	if s.Flags == nil {
+		// TODO: remove this fallback
+		s.Flags = pflag.NewFlagSet("", pflag.ContinueOnError)
+		if s.ParseFunc == nil {
+			s.ParseFunc = func() error { return s.Flags.Parse(os.Args[1:]) }
+		}
+	}
+
+	if s.NameCfg == nil {
+		s.NameCfg = DefaultFlagNameConfig()
+	}
+
+	if !s.flagsRegistered {
+		var ptyp reflect.Type
+		if s.ptrType == nil {
+			ptyp = t.Type()
+		} else {
+			ptyp = s.ptrType
+		}
+
+		if err := s.registerPFlags(reflect.Value{}, ptyp); err != nil {
+			return reflect.Value{}, err
+		}
+		s.flagsRegistered = true
+	}
+	if !s.Flags.Parsed() {
+		if err := s.parse(); err != nil {
+			return reflect.Value{}, err
+		}
+	}
+	var setErr error
+	val := reflect.New(t.Type())
+	s.Flags.Visit(func(f *pflag.Flag) {
+		fieldName, ok := s.flagFieldName[f.Name]
+		if !ok {
+			return
+		}
+
+		ffield := s.trnslVal.FieldByName(fieldName)
+		if !ffield.IsNil() {
+			// there's a 1:1 mapping between flags and field names so panic if
+			// this happens
+			panic(fmt.Errorf("Field name %s with flag %s is nil", fieldName, f.Name))
+		}
+
+		// We'll assume we're in a pointerified struct that matches
+		// what we expected before, here.
+		ptrVal := reflect.New(stripTypePtr(ffield.Type()))
+		if g, ok := s.flagValues[f.Name]; ok {
```

Comment:
Given that this `if` statement is the entire rest of the function-body, it would be cleaner to pull out the initializing statement into its own line and return early.

Reply:
Changed in pflag and flags. Looks much cleaner :)